README
libb64-2: RFC4648 Base64 Encoding/Decoding Routines
======================================

Overview:
--------
libb64-2 is a header library of ANSI C routines for fast encoding/decoding data
into and from a base64-encoded format. C++ wrappers are included, as well as
the source code for standalone encoding and decoding executables.

base64 consists of ASCII text, and is therefore a useful encoding for storing 
binary data in a text file, such as xml, or sending binary data over text-only
email.

References:
----------
* Wikipedia article:
	http://en.wikipedia.org/wiki/Base64
* base64, another implementation of a commandline en/decoder:
	http://www.fourmilab.ch/webtools/base64/

Why?
----
[by libb64 author Chris Venter]
I did this because I need an implementation of base64 encoding and decoding,
without any licensing problems. Most OS implementations are released under
either the GNU/GPL, or a BSD-variant, which is not what I require.

Also, the chance to actually use the co-routine implementation in code is rare,
and its use here is fitting. I couldn't pass up the chance.
For more information on this technique, see "Coroutines in C", by Simon Tatham,
which can be found online here: 
http://www.chiark.greenend.org.uk/~sgtatham/coroutines.html

base64 Commandline Use:
---------------
There is a new executable available, it is simply called base64.
It can encode and decode files, as instructed by the user.

To encode a file:
$ ./base64 -e filea fileb
fileb will now be the base64-encoded version of filea.

To decode a file:
$ ./base64 -d fileb filec
filec will now be identical to filea.

Programming:
-----------
The header library is written in C.  Simply include cencode.h and cdecode.h
in your C sources to get started.  The libary does not allocate memory.  You
provide the input and output buffers as pointers to char.  Remember that
Base64 encodes n octects in 4n/3 chars.

C++ wrappers are provided as well, so you don't have to get your hands
dirty. Encoding from standard input to standard output is as simple as

	#include <b64/encode.hpp>
	#include <iostream>
	int main()
	{
		base64::encoder E;
		E.encode(std::cin, std::cout);
		return 0;
	}

The encode method takes any istream and ostream, and uses the get() and put()
methods to read/write each octet.
	
The standalone executable, base64, demonstrates the use of the header library
found in include/b64.  The header library does not require linking, simply
include the encoder or decoder headers wherever you need them.  Include the .h
files in C code and the .hpp files in C++.  The .hpp files pull in the C headers
for you, too.

Implementation:
--------------
[by libb64 author Chris Venter]
It is DAMN fast, if I may say so myself. The C code uses a little trick which
has been used to implement coroutines, of which one can say that this
implementation is an example.

(To see how the libb64 codebase compares with some other BASE64 implementations
available, see the BENCHMARKS file)

The trick involves the fact that a switch-statement may legally cross into
sub-blocks. A very thorough and enlightening essay on co-routines in C, using
this method, can be found in the above mentioned "Coroutines in C", by Simon
Tatham: http://www.chiark.greenend.org.uk/~sgtatham/coroutines.html

For example, an RLE decompressing routine, adapted from the article:
1	static int STATE = 0;
2	static int len = 0, c;
3	switch (STATE)
4	{
5		while (1)
6		{
7			c = getchar();
8			if (c == EOF) return EOF;
9			if (c == 0xFF) {
10				len = getchar();
11				c = getchar();
12				while (len--)
13				{
14					STATE = 0;
15					return c;
16	case 0:
17				}
18			} else
19				STATE = 1;
20				return c;
21	case 1:
22			}
23		}
24	}

As can be seen from this example, a coroutine depends on a state variable,
which it sets directly before exiting (lines 14 and 119). The next time the
routine is entered, the switch moves control to the specific point directly
after the previous exit (lines 16 and 21).hands

(As an aside, in the mentioned article the combination of the top-level switch,
the various setting of the state, the return of a value, and the labelling of
the exit point is wrapped in #define macros, making the structure of the
routine even clearer.)

The obvious problem with any such routine is the static keyword.
Any static variables in a function spell doom for multithreaded applications.
Also, in situations where this coroutine is used by more than one other
coroutines, the consistency is disturbed.

What is needed is a structure for storing these variabled, which is passed to
the routine seperately. This obviously breaks the modularity of the function,
since now the caller has to worry about and care for the internal state of the
routine (the callee). This allows for a fast, multithreading-enabled
implementation, which may (obviously) be wrapped in a C++ object for ease of
use.

The base64 encoding and decoding functionality in this package is implemented
in exactly this way, providing both a high-speed high-maintanence C interface,
and a wrapped C++ which is low-maintanence and only slightly less performant.

Langley here: Notic that case statements can not only cross into sub-blocks,
but can jump past the initial checks of conditional loops to the first trip
through the loop starts on a given line.  Take a look at line 16 in the listing
above.  In spite of the condition's syntactic placement at the top of the while
loop, the 'len-- != 0' condition is checked at the bottom of the block, too. In
the co-routine model, this is acutally a do..while().  Neat, isn't it?

On the first call to the RLE decompression routine, the switch jumps to line
16. The inner while(len--) loop condition breaks on line 17, falls through to
the end of the 'while(1)' loop at line 23, then jumps back to the top of the
while(1) loop - which was skipped by the switch - and consumes the first cha-
racter of input.

Chris is correct when he says the C code is high-maintenance.  It's also quite
elegant.  The more you know...


CHANGELOG
libb64: Base64 Encoding/Decoding Routines
======================================

## Changelog ##

Version 2.0 Release
-------------------
Brought the library into compliance with RFC 4648.
http://tools.ietf.org/html/rfc4648
This implementation ignores CRLF and other newline sequences adjacent to the
data stream.
This implementation does not add newlines to the encoded data stream.
This implementation rejects encoded data streams that contain invalid
characters.  Decoding aborts and returns -1 (unless the called method returns
void).

Refactored the library into a "header library."  With a header library,
there's no object to link.  Simply include the header and use the functions.

Recreated the Visual Studio solution in VS C++ Express 2008 edition for
additional compatibility. (2010 edition automatically upgrades earlier solutions.

Placed the update library under the Apache License, Version 2.0.  The Creative
Commons retired the Public Domain Dedication and Certification in 2010.  Chose
the Apache License v2 over public domain dedication in order to ensure patent
and trademark licenses are freely available to users in addition to the code.

Version 1.2 Release
-------------------
Removed the b64dec, b64enc, encoder and decoder programs in favour of
a better example, called base64, which encodes and decodes 
depending on its arguments.

Created a solution for Microsoft Visual Studio C++ Express 2010
edition, which simply builds the base64 example as a console application.

Version 1.1 Release
-------------------
Modified encode.h to (correctly) read from the iostream argument,
instead of std::cin.
Thanks to Peter K. Lee for the heads-up.

No API changes.

Version 1.0 Release
-------------------
The current content is the changeset.


base64/base64.cpp
/*
base64.cc - c++ source to a base64 reference encoder and decoder

This is part of the libb64 project, and has been placed in the public domain.
For details, see http://sourceforge.net/projects/libb64
*/

#include <b64/encode.hpp>
#include <b64/decode.hpp>

#include <iostream>
#include <fstream>
#include <string>

#include <stdlib.h>

// Function which prints the usage of this executable
void usage()
{
    std::cerr<< \
        "base64: Encodes and Decodes files using base64\n" \
        "Usage: base64 [-e|-d] [input] [output]\n" \
        "   Where [-e] will encode the input file into the output file,\n" \
        "         [-d] will decode the input file into the output file, and\n" \
        "         [input] and [output] are the input and output files, respectively.\n";
}
// Function which prints the usage of this executable, plus a short message
void usage(const std::string& message)
{
    usage();
    std::cerr<<"Incorrect invocation of base64:\n";
    std::cerr<<message<<std::endl;
}

int main(int argc, char** argv)
{
    // Quick check for valid arguments
    if (argc == 1)
    {
        usage();
        exit(-1);
    }
    if (argc != 4)
    {
        usage("Wrong number of arguments!");
        exit(-1);
    }
    
    // So far so good; try to open the input file
    std::string input = argv[2];
    // Note that we have to open the input in binary mode.
    // This is due to some operating systems not using binary mode by default.
    // Since we will most likely be dealing with binary files when encoding, we
    // have to be able to deal with zeros (and other invalid chars) in the input stream.
    std::ifstream instream(input.c_str(), std::ios_base::in | std::ios_base::binary);
    if (!instream.is_open())
    {
        usage("Could not open input file!");
        exit(-1);
    }
    
    // Now try to open the output file
    std::string output = argv[3];
    // Again, note that we have to open the ouput in binary mode.
    // Similiarly, we will most likely need to deal with zeros in the output stream when we
    // are decoding, and the output stream has to be able to use these invalid text chars.
    std::ofstream outstream(output.c_str(), std::ios_base::out | std::ios_base::binary);
    if (!outstream.is_open())
    {
        usage("Could not open output file!");
        exit(-1);
    }

    // determine whether we need to encode or decode:
    std::string choice = argv[1];
    if (choice == "-d")
    {
        base64::decoder D;
        D.decode(instream, outstream);
    }
    else if (choice == "-e")
    {
        base64::encoder E;
        E.encode(instream, outstream);
    }
    else
    {
        std::cout<<"["<<choice<<"]"<<std::endl;
        usage("Please specify -d or -e as first argument!");
    }

    return 0;
}



include/b64/cencode.h
/*
cencode.h - c header for a base64 encoding algorithm

Copyright 2012 Joseph R. Langley

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Derived from the libb64 project written and dedicated to the public
domain by Chris Venter.
http://libb64.sourceforge.net/

*/

#ifndef BASE64_CENCODE_H
#define BASE64_CENCODE_H

/*
 * Declarations
 */

typedef enum
{
    step_A, /* 1st octet of a 3-octet block */
    step_B, /* 2nd octet of a 3-octet block */
    step_C  /* 3rd octet of a 3-octet block */
} base64_encodestep;

typedef struct
{
    base64_encodestep step;
    char result;
} base64_encodestate;

void base64_init_encodestate( base64_encodestate* state_in );

char base64_encode_value( char value_in );

int base64_encode_block( const char* plaintext_in, 
                         int length_in,
                         char* code_out,
                         base64_encodestate* state_in );

int base64_encode_blockend( char* code_out,
                            base64_encodestate* state_in );

/*
 * Definitions
 */ 

void base64_init_encodestate( base64_encodestate* state_in ) {
    state_in->step = step_A;
    state_in->result = 0;
}

char base64_encode_value( char value_in ) {
    static const char encoding[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    static const char encoding_size = sizeof(encoding);
    if (value_in >= encoding_size) {
        return '=';
    }
    return encoding[(int)value_in];
}

int base64_encode_block( const char* plaintext_in,
                         int length_in,
                         char* code_out,
                         base64_encodestate* state_in ) {
    const char* plainchar = plaintext_in;
    const char* const plaintextend = plaintext_in + length_in;
    char* codechar = code_out;
    char result;
    char fragment;
    
    result = state_in->result;
    
    switch (state_in->step)
    {
        while (1)
        {
    case step_A:
            if (plainchar == plaintextend)
            {
                state_in->result = result;
                state_in->step = step_A;
                return codechar - code_out;
            }
            fragment = *plainchar++;
            result = (fragment & 0x0fc) >> 2;
            *codechar++ = base64_encode_value(result);
            result = (fragment & 0x003) << 4;
    case step_B:
            if (plainchar == plaintextend)
            {
                state_in->result = result;
                state_in->step = step_B;
                return codechar - code_out;
            }
            fragment = *plainchar++;
            result |= (fragment & 0x0f0) >> 4;
            *codechar++ = base64_encode_value(result);
            result = (fragment & 0x00f) << 2;
    case step_C:
            if (plainchar == plaintextend)
            {
                state_in->result = result;
                state_in->step = step_C;
                return codechar - code_out;
            }
            fragment = *plainchar++;
            result |= (fragment & 0x0c0) >> 6;
            *codechar++ = base64_encode_value(result);
            result  = (fragment & 0x03f) >> 0;
            *codechar++ = base64_encode_value(result);
        }
    }
    /* control should not reach here */
    return codechar - code_out;
}

int base64_encode_blockend( char* code_out,
                            base64_encodestate* state_in ) {
    char* codechar = code_out;
    
    switch (state_in->step)
    {
    case step_B:
        *codechar++ = base64_encode_value(state_in->result);
        *codechar++ = '=';
        *codechar++ = '=';
        break;
    case step_C:
        *codechar++ = base64_encode_value(state_in->result);
        *codechar++ = '=';
        break;
    case step_A:
        break;
    }
    
    return codechar - code_out;
}

#endif /* BASE64_CENCODE_H */


include/b64/decode.hpp
// :mode=c++:
/*
decode.h - c++ wrapper for a base64 decoding algorithm

Copyright 2012 Joseph R. Langley

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Derived from the libb64 project written and dedicated to the public
domain by Chris Venter.
http://libb64.sourceforge.net/

*/
#ifndef BASE64_DECODE_H
#define BASE64_DECODE_H

#include <iostream>

namespace base64
{
    extern "C"
    {
        #include "cdecode.h"
    }

    struct decoder
    {
        base64_decodestate _state;
        int _buffersize;

        decoder(int buffersize_in = BUFFERSIZE)
        : _buffersize(buffersize_in)
        {}

        int decode(char value_in)
        {
            return base64_decode_value(value_in);
        }

        int decode(const char* code_in, const int length_in, char* plaintext_out)
        {
            return base64_decode_block(code_in, length_in, plaintext_out, &_state);
        }

        void decode(std::istream& istream_in, std::ostream& ostream_in)
        {
            base64_init_decodestate(&_state);
            //
            const int N = _buffersize;
            char* code = new char[N];
            char* plaintext = new char[N];
            int codelength;
            int plainlength;

            do
            {
                istream_in.read((char*)code, N);
                codelength = istream_in.gcount();
                if( (plainlength = decode(code, codelength, plaintext)) < 0 ) {
                    std::cerr << "Breaking because of invalid encoded characters." << std::endl;
                    break;
                }
                ostream_in.write((const char*)plaintext, plainlength);
            }
            while (istream_in.good() && codelength > 0);
            //
            base64_init_decodestate(&_state);

            delete [] code;
            delete [] plaintext;
        }
    };

} // namespace base64



#endif // BASE64_DECODE_H



include/b64/encode.hpp
/*
encode.hpp - c++ wrapper for a base64 encoding algorithm

Copyright 2012 Joseph R. Langley

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Derived from the libb64 project written and dedicated to the public
domain by Chris Venter.
http://libb64.sourceforge.net/
 
*/
#ifndef BASE64_ENCODE_H
#define BASE64_ENCODE_H

#include <iostream>

namespace base64
{
    extern "C" 
    {
        #include "cencode.h"
    }

    struct encoder
    {
        base64_encodestate _state;
        int _buffersize;

        encoder(int buffersize_in = BUFFERSIZE)
        : _buffersize(buffersize_in)
        {}

        int encode(char value_in)
        {
            return base64_encode_value(value_in);
        }

        int encode(const char* code_in, const int length_in, char* plaintext_out)
        {
            return base64_encode_block(code_in, length_in, plaintext_out, &_state);
        }

        int encode_end(char* plaintext_out)
        {
            return base64_encode_blockend(plaintext_out, &_state);
        }

        void encode(std::istream& istream_in, std::ostream& ostream_in)
        {
            base64_init_encodestate(&_state);
            //
            const int N = _buffersize;
            char* plaintext = new char[N];
            char* code = new char[2*N];
            int plainlength;
            int codelength;

            do
            {
                istream_in.read(plaintext, N);
                plainlength = istream_in.gcount();
                //
                codelength = encode(plaintext, plainlength, code);
                ostream_in.write(code, codelength);
            }
            while (istream_in.good() && plainlength > 0);

            codelength = encode_end(code);
            ostream_in.write(code, codelength);
            //
            base64_init_encodestate(&_state);

            delete [] code;
            delete [] plaintext;
        }
    };

} // namespace base64

#endif // BASE64_ENCODE_H



include/b64/cdecode.h
/*
cdecode.h - c header for a base64 decoding algorithm

Copyright 2012 Joseph R. Langley

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Derived from the libb64 project written and dedicated to the public
domain by Chris Venter.
http://libb64.sourceforge.net/

*/

#ifndef BASE64_CDECODE_H
#define BASE64_CDECODE_H

/* 
 * Declarations 
 */

typedef enum
{
    step_a, /* 1st encoded char of a 3-octet block */
    step_b, /* 2nd encoded char of a 3-octet block */
    step_c, /* 3rd encoded char of a 3-octet block */
    step_d  /* 4th encoded char of a 3-octet block */
} base64_decodestep;

typedef struct
{
    base64_decodestep step;
    char plainchar;
} base64_decodestate;

void base64_init_decodestate( base64_decodestate* state_in );

int base64_decode_value( char value_in );

int base64_decode_block( const char* code_in, 
                         const int length_in,
                         char* plaintext_out,
                         base64_decodestate* state_in );

/*
 * Definitions
 */
 
int base64_decode_value( char value_in ) {
    /*
     * Some beautiful magic is happening here.  We're trying to decode
     * characters from ASCII encoding into binary according to an en-
     * coding sequence that didn't translate linearly from binary to 
     * ASCII code points.  Fun, huh?
     *
     * Here's what you're seeing:
     * The incoming value gets reduced by 43.  Why?  Because that's the 
     * ASCII code point for '+', the smallest-valued character in the
     * Base64 endcoding table.  The decoding matrix below contains the
     * decimal values for all the code points between ASCII '+' and 'z'.
     * Those that aren't valid Base64 characters return -1 so we can
     * detect the input error in the base64_decode_block() function.
     *
     * Finally, the -2 at offset 18 in the decoding array marks the
     * position of ASCII '=', the padding character specified by RFC4648.
     * Finding that means we're in the padding at the end of an encoded
     * stream.
     */
    static const char decoding[] = {
        62,-1,-1,-1,63,52,53,54,55,56,57,58,59,60,
        61,-1,-1,-1,-2,-1,-1,-1, 0, 1, 2, 3, 4, 5,
         6, 7, 8, 9,10,11,12,13,14,15,16,17,18,19,
        20,21,22,23,24,25,-1,-1,-1,-1,-1,-1,26,27,
        28,29,30,31,32,33,34,35,36,37,38,39,40,41,
        42,43,44,45,46,47,48,49,50,51
    };
    static const char decoding_size = sizeof(decoding);
    value_in -= 43; /* ASCII '+' == 43 */
    if (value_in < 0 || value_in > decoding_size) {
        return -1; /* This handles any input less than '+' or greater */
                   /* than 'z'.                                       */
    }
    return decoding[(int)value_in];
}

void base64_init_decodestate( base64_decodestate* state_in ) {
    state_in->step = step_a;
    state_in->plainchar = 0;
    return;
}

int base64_decode_block( const char* code_in,
                         const int length_in,
                         char* plaintext_out,
                         base64_decodestate* state_in ) {

    const char* codechar = code_in;
    char* plainchar = plaintext_out;
    char fragment;
    
    *plainchar = state_in->plainchar;
    
    switch (state_in->step)
    {
        while (1)
        {
    case step_a:
            do {
                /* Handle the end of an encoded block. */
                /* Note: codechar points at the byte   */
                /* after the end of code_in.           */
                if (codechar == code_in+length_in)
                {
                    state_in->step = step_a;
                    state_in->plainchar = *plainchar;
                    /* Return the number of decoded bytes */
                    return plainchar - plaintext_out;
                }
                /* Handle any encoded character */
                fragment = (char)base64_decode_value(*codechar++);
                if( fragment == -1 ) goto error; /* Invalid encoding found   */
            } while( fragment < -1 ); /* Loop through the padding at the end */
                                      /* of an encoded block.                */
            /* Get the six most significant bits of the first encoded octet  */
            *plainchar    = (fragment & 0x03f) << 2;
    case step_b:
            do {
                if (codechar == code_in+length_in)
                {
                    state_in->step = step_b;
                    state_in->plainchar = *plainchar;
                    return plainchar - plaintext_out;
                }
                fragment = (char)base64_decode_value(*codechar++);
                if( fragment == -1 ) goto error;
            } while( fragment < -1 );
            /* Get the two least significant bits of the first encoded octet */
            *plainchar++ |= (fragment & 0x030) >> 4;
            /* Get the four msb of the second encoded octet */
            *plainchar    = (fragment & 0x00f) << 4;
    case step_c:
            do {
                if (codechar == code_in+length_in)
                {
                    state_in->step = step_c;
                    state_in->plainchar = *plainchar;
                    return plainchar - plaintext_out;
                }
                fragment = (char)base64_decode_value(*codechar++);
                if( fragment == -1 ) goto error;
            } while( fragment < -1 );
            /* Get the four lsb of the second encoded octet */
            *plainchar++ |= (fragment & 0x03c) >> 2;
            /* Get the two msb of the third encoded octet */
            *plainchar    = (fragment & 0x003) << 6;
    case step_d:
            do {
                if (codechar == code_in+length_in)
                {
                    state_in->step = step_d;
                    state_in->plainchar = *plainchar;
                    return plainchar - plaintext_out;
                }
                fragment = (char)base64_decode_value(*codechar++);
                if( fragment == -1 ) goto error;
            } while( fragment < -1 );
            /* Ger the four lsb of the third encoded octet */
            *plainchar++   |= (fragment & 0x03f);
        }
error:
        /* control comes here when invalid characters occur in the data. */
        return -1;
    }
    /* control should not reach here */
    return plainchar - plaintext_out;

}

#endif /* BASE64_CDECODE_H */
